app/vmalert/manager.go

```diff
@@ -93,31 +93,53 @@ func (m *manager) update(ctx context.Context, path []string, validateTpl, valida
 		groupsRegistry[ng.ID()] = ng
 	}
 
+	type updateItem struct {
+		old *Group
+		new *Group
+	}
+	var toUpdate []updateItem
+
 	m.groupsMu.Lock()
 	for _, og := range m.groups {
 		ng, ok := groupsRegistry[og.ID()]
 		if !ok {
-			// old group is not present in new list
-			// and must be stopped and deleted
+			// old group is not present in new list,
+			// so must be stopped and deleted
 			og.close()
 			delete(m.groups, og.ID())
 			og = nil
 			continue
 		}
-		og.updateCh <- ng
 		delete(groupsRegistry, ng.ID())
+		if og.Checksum != ng.Checksum {
+			toUpdate = append(toUpdate, updateItem{old: og, new: ng})
+		}
 	}
-
 	for _, ng := range groupsRegistry {
 		m.startGroup(ctx, ng, restore)
 	}
 	m.groupsMu.Unlock()
+
+	if len(toUpdate) > 0 {
```

Comment:
do we need to return nil here if toUpdate lentgh equals zero

Reply:
No. Configuration reload may contain following changes:
1. groups to remove
2. groups to add
3. groups to update

All of the listed changes are optional and config reload may not bring any change at all.

Reply:
Do you mean remove extra indentation by the following code?
```
if len(toUpdate) < 1 {
		return nil
	}
	var wg sync.WaitGroup
	for _, item := range toUpdate {
		wg.Add(1)
		go func(old *Group, new *Group) {
			old.updateCh <- new
			wg.Done()
		}(item.old, item.new)
	}
	wg.Wait()
	return nil
```